strings/bitap_string_match.py

```diff
@@ -0,0 +1,79 @@
+"""
+Bitap exact string matching
+https://en.wikipedia.org/wiki/Bitap_algorithm
+
+Searches for a pattern inside text, and returns the index of the first occurrence
+of the pattern. Both text and pattern consist of lowercase alphabetical characters only.
+
+Complexity: O(m*n)
+    n = length of text
+    m = length of pattern
+
+Python doctests can be run using this command:
+python3 -m doctest -v bitap_string_match.py
+"""
+
+
+def bitap_string_match(text: str, pattern: str) -> int:
+    """
+    Retrieves the index of the first occurrence of pattern in text.
+
+    Args:
+        text: A string consisting only of lowercase alphabetical characters.
+        pattern: A string consisting only of lowercase alphabetical characters.
+
+    Returns:
+        int: The index where pattern first occurs. Return -1  if not found.
+
+    >>> bitap_string_match('abdabababc', 'ababc')
+    5
+    >>> bitap_string_match('aaaaaaaaaaaaaaaaaa', 'a')
+    0
+    >>> bitap_string_match('zxywsijdfosdfnso', 'zxywsijdfosdfnso')
+    0
+    >>> bitap_string_match('abdabababc', '')
+    0
+    >>> bitap_string_match('abdabababc', 'c')
+    9
+    >>> bitap_string_match('abdabababc', 'fofosdfo')
+    -1
+    >>> bitap_string_match('abdab', 'fofosdfo')
+    -1
+    """
+    m: int = len(pattern)
+    if m == 0:
+        return 0
```

Comment:
All falsey values for `pattern` are equally dangerous as `""`.

Python and mypy are both smart enough to figure out without the type hint that if `len(pattern)` is assigned to `m` then `m` is an int.

```suggestion
    if not pattern:
        return 0
    m = len(pattern)
```